feat: Add [account/user-partition] resource limit

huerni opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Support for partition-specific resource limits for accounts and users; ModifyAccount now accepts a partition target and account/user queries expose per-partition limits.

* **Behavior Changes**
  * Scheduling and resource accounting now enforce both QoS and partition-scoped limits, affecting submit/run decisions.

* **Bug Fixes / Docs**
  * Added new error codes and pending-reason entries documenting partition and QoS limit failures.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

huerni commented:
merged after #528

huerni commented:
@coderabbitai full review

L-Xiafeng commented on src/CraneCtld/AccountMetaContainer.h:
```diff
@@ -21,33 +21,111 @@
-  using QosToResourceMap = std::unordered_map<std::string,  // qos_name
-                                              QosResource>;
+  using ResourceMetaMap = phmap::parallel_flat_hash_map<
+      std::string, MetaResourceStat,
```
Account和User单独写两个Map类型吧？UserToTaskNumMap 可以合并下，然后在在这里加一下qos_id_t account_id_t,user_id_t 都是string对应entity的name

L-Xiafeng commented on src/CraneCtld/AccountMetaContainer.h:
```diff
@@ -21,33 +21,111 @@
-  CraneErrCode TryMallocQosResource(TaskInCtld& task);
+  CraneErrCode TryMallocMetaSubmitResource(TaskInCtld& task);
+
+  void MallocMetaSubmitResource(const TaskInCtld& task);
```
MallocResourceOnSubmitted

L-Xiafeng commented on src/CraneCtld/AccountMetaContainer.h:
```diff
@@ -21,33 +21,111 @@
+
+  void MallocMetaSubmitResource(const TaskInCtld& task);
+
+  void MallocMetaResourceToRecoveredRunningTask(TaskInCtld& task);
```
MallocResourceOnRecover

L-Xiafeng commented on src/CraneCtld/AccountMetaContainer.h:
```diff
@@ -21,33 +21,111 @@
+
+  std::optional<std::string> CheckMetaResource(const TaskInCtld& task);
+
+  void MallocMetaResource(const TaskInCtld& task);
```
MallocMetaResourceOnSched

huerni committed:
feat: account partition limit - keep only proto and AccountManager changes on top of latest master

huerni committed:
feat: Add Suspend and Resume job requests and update TerminateStepRequest with terminate source

huerni committed:
style: Improve code formatting and comments in proto files for consistency

huerni committed:
feat: Update AccountManager and related components to support job-based resource limits and modify account operations

huerni committed:
Refactor resource management in AccountMetaContainer

- Introduced MetaResourceStat to track per-user and per-account resource usage across QoS and partition dimensions.
- Updated method names for clarity, changing "Qos" to "Meta" in resource management functions to reflect the new structure.
- Replaced old resource checking and allocation methods with new implementations that utilize the MetaResourceStat structure.
- Adjusted JobScheduler to use the new resource management methods, ensuring compatibility with the updated resource allocation logic.

huerni committed:
feat: Refactor partition limit checks in AccountMetaContainer for improved clarity and efficiency

huerni committed:
feat: Add IsUnlimitedTres_ check for resource limits in AccountMetaContainer

huerni committed:
refactor: Remove commented-out section for clarity in AccountMetaContainer

huerni committed:
feat: Implement partition resource limit handling in MongodbClient and update gRPC service methods

huerni committed:
fix: Add missing newline at end of file in proto definitions

huerni committed:
refactor: Update resource limit handling in AccountManager and DbClient for improved clarity and efficiency

huerni committed:
refactor: Simplify resource limit checks in AccountMetaContainer and enhance partition limit handling

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
refactor: Enhance CheckTres_ method to include prefix for clearer error messages

huerni committed:
refactor: Enhance resource allocation checks in TryMallocMetaSubmitResource method

huerni committed:
refactor: Update error codes and messages for partition resource limits in PublicDefs and AccountMetaContainer

huerni committed:
refactor: Add detailed resource limit pending reasons for QoS and Partition in documentation

huerni committed:
refactor: Improve error handling and resource limit checks in AccountManager and AccountMetaContainer

huerni committed:
refactor: Rename error code for missing partition entry in pending reasons and AccountMetaContainer

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
refactor: Enhance documentation for resource management functions and add static assertion for error code consistency

huerni committed:
docs: Add Resource Limit Configuration Guide and update deployment index

- Introduced a new guide for configuring QoS and partition-level resource limits for accounts and users.
- Updated the deployment index to include a link to the new Resource Limit Configuration Guide.
- Added the Resource Limit Guide to the navigation in mkdocs.yaml for better accessibility.

huerni committed:
feat: Add partition resource limits and mapping for accounts and users

huerni committed:
merge master

NamelessOIer commented on src/CraneCtld/Accounting/AccountMetaContainer.cpp:
```diff
@@ -264,93 +263,280 @@ void AccountMetaContainer::DeleteQosMeta(const std::string& qos) {
+  // ---- User entity ----
+  {
+    const PartitionResourceLimit* user_part_limit = nullptr;
+    auto user_ptr = g_account_manager->GetExistedUserInfo(job.Username());
```
这个锁外面拿过了吧

huerni replied:
确实是，rebase 错了

huerni committed:
feat: Add account/user partition limits and update error string array size

huerni committed:
feat: Enhance resource submission checks by incorporating user context and account mapping

huerni committed:
feat: Update partition submit limit checks to incorporate QoS parameters

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
feat: Update resource limit checks to use QoS GRES validation

huerni committed:
fix

huerni committed:
feat: Enhance resource management by adding count parameter to submission functions

huerni committed:
refactor: Update error string array to use dynamic size based on ErrCode_ARRAYSIZE

github-actions[bot] committed:
style: auto format with clang-format.

huerni committed:
feat: Update account and user modification operations to join value lists as strings

huerni committed:
feat: Enhance resource tracking by adding account-level partition resource mapping

huerni committed:
feat: Improve partition limit checks with detailed logging for user and account submissions

github-actions[bot] committed:
style: auto format with clang-format.

NamelessOIer commented on src/CraneCtld/Accounting/AccountMetaContainer.cpp:
```diff
@@ -298,250 +323,695 @@ void AccountMetaContainer::DeleteQosMeta(const std::string& qos) {
+      }
+      if (qos.max_submit_jobs_per_user ==
+          std::numeric_limits<decltype(qos.max_submit_jobs_per_user)>::max()) {
+        if (user_part_limit->max_submit_jobs == 0) {
```
这里 == 0 校验对 job array 有问题，一个 job array 可能有多个 job，应当实际比大小

huerni replied:
<img width="595" height="472" alt="image" src="https://github.com/user-attachments/assets/eda0667a-6d27-41eb-857d-e082a654ca2c" />

我看master，第一次提交qos都会有这个问题？然后job.req_rotal_res_view是一个job的资源还是所有job的资源？如果是一个job的资源，也是会有问题。

huerni replied:
如果确实都有问题，我这里可以先把partition的改了，partition合入后，再修改qos的

NamelessOIer replied:
应该是都有这个问题，但应该只会动相关的判断逻辑？可以直接在这个分支一块修了吧，其他没什么问题

huerni replied:
测试那边可能会有其他用例改动，和周老师商量了，后续更改吧

NamelessOIer commented on src/CraneCtld/Accounting/AccountMetaContainer.cpp:
```diff
@@ -298,250 +323,695 @@ void AccountMetaContainer::DeleteQosMeta(const std::string& qos) {
+  }
+
+  if (!CheckGres_(resource_req.GetGresMap(), resource_total.GetGresMap()))
+    return std::unexpected("QosGresResourceLimit");
```
这个地方现在有 partiton 不一定是 qos 了

NamelessOIer commented on docs/en/reference/error_code.md:
```diff
@@ -79,3 +79,35 @@
+| 103 | `ERR_MAX_TRES_PER_ACCOUNT_BEYOND` | Current account's TRES usage reached QoS limit | Contact administrator to raise QoS account TRES limit | The tres of jobs for the current account has reached the limit of this QOS |
+| 104 | `ERR_TRES_PER_JOB_BEYOND` | Single job's TRES exceeds QoS limit | Reduce job resource request or contact administrator to raise QoS limit | The tres of jobs has reached the limit of this QOS |
+| 105 | `ERR_INVALID_DEADLINE` | Invalid deadline time (should be later than submit time) | Set a deadline time later than the current time | Invalid deadline time(should be later than submit time) |
+| 106 | `ERR_PARTITION_TRES_PER_JOB_BEYOND` | Single job's TRES exceeds Partition's `max_tres_per_job` limit | Reduce job resource request or contact administrator to raise Partition limit | Resource (TRES) per job exceeds the partition limit |
```
ERR_PMIX_ERR 漏了

huerni commented:
@NamelessOIer 还有其他问题吗？

huerni committed:
feat: Add partition resource limit options and enhance error handling in account management

github-actions[bot] committed:
style: auto format with clang-format.

NamelessOIer commented on src/CraneCtld/Accounting/AccountMetaContainer.cpp:
```diff
@@ -723,7 +723,7 @@ CraneErrCode AccountMetaContainer::CheckSubmitLimits_(
       if (qos.max_submit_jobs_per_user ==
           std::numeric_limits<decltype(qos.max_submit_jobs_per_user)>::max()) {
-        if (user_part_limit->max_submit_jobs == 0) {
+        if (user_part_limit->max_submit_jobs == count-1) {
```
这里应该是 <= 吧，或者统一风格应该是 
`count > user_part_limit->max_submit_jobs `

huerni replied:
修了

huerni committed:
fix
